File: findings.java
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.*;
import com.fasterxml.jackson.databind.node.ObjectNode;
import okhttp3.*;
import org.apache.commons.cli.*;
import org.apache.commons.csv.*;

import java.io.*;
import java.net.URLEncoder;
import java.nio.file.*;
import java.time.*;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.concurrent.TimeUnit;

public class NonameFindingsFetcher {

    // Config - replace these with your real values or read from env/config file
    private static final String CLIENT_ID = "your_client_id";
    private static final String CLIENT_SECRET = "your_client_secret";
    private static final String HOST = "https://wellsfargobank.nonamesec.com";
    private static final String AUTH_URL = HOST + "/auth/token";
    private static final String FINDINGS_API = HOST + "/api/v4/findings";
    private static final String API_METADATA_ENDPOINT = HOST + "/api/v3/apis/";

    private static final Path TOKEN_CACHE_FILE = Paths.get(System.getProperty("java.io.tmpdir"), "token_cache.json");
    private static final int TOKEN_EXPIRY_HOURS = 8;
    private static final int LIMIT = 50;

    private static final Path CSV_PATH = Paths.get(System.getProperty("java.io.tmpdir"), "findings_report.csv");
    private static final Path JSON_PATH = Paths.get(System.getProperty("java.io.tmpdir"), "findings_report.json");

    private static OkHttpClient client = new OkHttpClient.Builder()
            .callTimeout(30, TimeUnit.SECONDS)
            .build();

    private static ObjectMapper mapper = new ObjectMapper()
            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);

    private static String bearerToken = null;

    // Fields to request from findings API
    private static final List<String> RETURN_FIELDS = Arrays.asList(
            "id", "title", "url", "typeId", "apiId", "module", "host", "path", "method",
            "resourceGroupName", "status", "severity", "owaspTags", "complianceFrameworkTags",
            "vulnerabilityFrameworkTags", "detectionTime", "lastUpdate", "triggeredOn",
            "description", "impact", "remediation", "investigate", "comments", "tickets",
            "externalTickets", "evidence", "source", "hasRelatedIncidents", "tagsIds",
            "relatedApiIds"
    );

    public static void main(String[] args) throws Exception {
        Options options = new Options();
        options.addOption("s", "start", true, "Start ISO timestamp (UTC)");
        options.addOption("e", "end", true, "End ISO timestamp (UTC)");
        options.addOption("h", "hours", true, "Hours back if no start/end");
        options.addOption(null, "include-api-metadata", false, "Include API metadata enrichment");

        CommandLineParser parser = new DefaultParser();
        CommandLine cmd = parser.parse(options, args);

        String start = cmd.getOptionValue("start");
        String end = cmd.getOptionValue("end");
        int hoursBack = Integer.parseInt(cmd.getOptionValue("hours", "24"));
        boolean includeApiMetadata = cmd.hasOption("include-api-metadata");

        bearerToken = getCachedToken();
        if (bearerToken == null) {
            bearerToken = fetchBearerToken();
            if (bearerToken == null) {
                System.err.println("❌ Failed to authenticate.");
                System.exit(1);
            }
        }

        String[] timeWindow = getTimeWindow(start, end, hoursBack);
        start = timeWindow[0];
        end = timeWindow[1];

        System.out.printf("📅 Fetching findings from: %s to %s%n", start, end);

        List<Map<String, Object>> rawFindings = fetchFindings(start, end);

        List<Map<String, Object>> outputFindings;
        if (includeApiMetadata) {
            outputFindings = enrichFindingsWithApiMetadata(rawFindings);
        } else {
            outputFindings = rawFindings;
        }

        writeFindingsToJson(outputFindings, JSON_PATH);
        writeFindingsToCsv(outputFindings, CSV_PATH);

        System.out.printf("📊 Total findings: %d%n", outputFindings.size());
    }

    private static String getCachedToken() {
        if (Files.exists(TOKEN_CACHE_FILE)) {
            try {
                byte[] bytes = Files.readAllBytes(TOKEN_CACHE_FILE);
                JsonNode root = mapper.readTree(bytes);
                String token = root.path("accessToken").asText(null);
                String timestampStr = root.path("timestamp").asText(null);
                if (token != null && timestampStr != null) {
                    Instant tokenTime = Instant.parse(timestampStr);
                    Instant now = Instant.now();
                    if (Duration.between(tokenTime, now).toHours() < TOKEN_EXPIRY_HOURS) {
                        System.out.println("🔐 Using cached token.");
                        return token;
                    }
                }
            } catch (IOException e) {
                System.err.println("⚠️ Failed to read or parse cached token: " + e.getMessage());
            }
        }
        return null;
    }

    private static void cacheToken(String token) {
        ObjectNode root = mapper.createObjectNode();
        root.put("accessToken", token);
        root.put("timestamp", Instant.now().toString());
        try {
            mapper.writerWithDefaultPrettyPrinter().writeValue(TOKEN_CACHE_FILE.toFile(), root);
        } catch (IOException e) {
            System.err.println("⚠️ Failed to write cached token: " + e.getMessage());
        }
    }

    private static String fetchBearerToken() {
        Map<String, String> payload = new HashMap<>();
        payload.put("client_id", CLIENT_ID);
        payload.put("client_secret", CLIENT_SECRET);
        String jsonPayload;
        try {
            jsonPayload = mapper.writeValueAsString(payload);
        } catch (JsonProcessingException e) {
            System.err.println("❌ Failed to serialize auth payload: " + e.getMessage());
            return null;
        }

        RequestBody body = RequestBody.create(jsonPayload, MediaType.get("application/json"));

        Request request = new Request.Builder()
                .url(AUTH_URL)
                .post(body)
                .build();

        try (Response response = client.newCall(request).execute()) {
            if (response.isSuccessful() && response.body() != null) {
                JsonNode json = mapper.readTree(response.body().string());
                String token = json.path("accessToken").asText(null);
                if (token != null) {
                    cacheToken(token);
                    return token;
                }
            } else {
                System.err.printf("❌ Token fetch failed: %d - %s%n", response.code(), response.message());
            }
        } catch (IOException e) {
            System.err.println("❌ Auth error: " + e.getMessage());
        }
        return null;
    }

    private static String[] getTimeWindow(String startArg, String endArg, int hoursBack) {
        Instant now = Instant.now();
        if (startArg != null && endArg != null) {
            return new String[]{startArg, endArg};
        }
        Instant start = now.minus(Duration.ofHours(hoursBack));
        Instant end = now;
        return new String[]{start.toString(), end.toString()};
    }

    private static List<Map<String, Object>> fetchFindings(String startTime, String endTime) {
        List<Map<String, Object>> findings = new ArrayList<>();
        int offset = 0;

        while (true) {
            HttpUrl.Builder urlBuilder = HttpUrl.parse(FINDINGS_API).newBuilder();
            urlBuilder.addQueryParameter("sortDesc", "true");
            urlBuilder.addQueryParameter("limit", String.valueOf(LIMIT));
            urlBuilder.addQueryParameter("offset", String.valueOf(offset));
            urlBuilder.addQueryParameter("detectionStartDate", startTime);
            urlBuilder.addQueryParameter("detectionEndDate", endTime);
            urlBuilder.addQueryParameter("lastUpdateStartDate", startTime);
            urlBuilder.addQueryParameter("lastUpdateEndDate", endTime);
            for (String field : RETURN_FIELDS) {
                urlBuilder.addQueryParameter("returnFields", field);
            }

            Request request = new Request.Builder()
                    .url(urlBuilder.build())
                    .header("Authorization", "Bearer " + bearerToken)
                    .header("Accept", "application/json")
                    .build();

            Response response = executeRequestWithRetry(request);

            if (response == null) break;

            try (ResponseBody body = response.body()) {
                if (body == null) break;
                JsonNode root = mapper.readTree(body.string());
                JsonNode entities = root.path("entities");
                if (entities.isArray()) {
                    for (JsonNode entity : entities) {
                        Map<String, Object> map = mapper.convertValue(entity, new TypeReference<Map<String, Object>>() {});
                        findings.add(map);
                    }
                    System.out.printf("📥 Retrieved %d findings (offset %d)%n", entities.size(), offset);
                }
                boolean more = root.path("moreEntities").asBoolean(false);
                if (!more) break;
            } catch (IOException e) {
                System.err.println("❌ Failed to parse findings response: " + e.getMessage());
                break;
            }

            offset += LIMIT;
        }
        return findings;
    }

    private static Response executeRequestWithRetry(Request request) {
        int retries = 5;
        for (int attempt = 0; attempt < retries; attempt++) {
            try {
                Response response = client.newCall(request).execute();
                if (response.code() == 401) {
                    // Token expired - refresh
                    System.out.println("🔁 Token expired. Refreshing.");
                    bearerToken = fetchBearerToken();
                    if (bearerToken == null) {
                        System.err.println("❌ Failed to refresh token.");
                        return null;
                    }
                    // Rebuild request with new token
                    request = request.newBuilder()
                            .header("Authorization", "Bearer " + bearerToken)
                            .build();
                    Thread.sleep(1500);
                    continue;
                }
                if (response.isSuccessful()) {
                    return response;
                } else {
                    System.err.printf("❗ Request failed (%d): %s%n", response.code(), response.message());
                }
                response.close();
            } catch (IOException | InterruptedException e) {
                System.err.println("❌ Network error: " + e.getMessage());
            }
            try {
                Thread.sleep((long) (Math.pow(2, attempt) * 1000 + Math.random() * 1000));
            } catch (InterruptedException ignored) {
            }
        }
        return null;
    }

    private static List<Map<String, Object>> enrichFindingsWithApiMetadata(List<Map<String, Object>> rawFindings) {
        List<Map<String, Object>> enriched = new ArrayList<>();
        Set<String> seenApiIds = new HashSet<>();

        for (Map<String, Object> finding : rawFindings) {
            Set<String> apiIds = new HashSet<>();
            if (finding.containsKey("apiId") && finding.get("apiId") != null) {
                apiIds.add(finding.get("apiId").toString());
            }
            if (finding.containsKey("relatedApiIds") && finding.get("relatedApiIds") instanceof List) {
                List<?> related = (List<?>) finding.get("relatedApiIds");
                for (Object o : related) {
                    if (o != null) apiIds.add(o.toString());
                }
            }

            List<Map<String, Object>> apiDetails = new ArrayList<>();
            for (String apiId : apiIds) {
                if (!seenApiIds.contains(apiId)) {
                    Map<String, Object> metadata = fetchApiMetadata(apiId);
                    if (metadata != null) {
                        apiDetails.add(metadata);
                    }
                    seenApiIds.add(apiId);
                }
            }
            finding.put("apiDetails", apiDetails);
            enriched.add(finding);
        }
        return enriched;
    }

    private static Map<String, Object> fetchApiMetadata(String apiId) {
        String url = API_METADATA_ENDPOINT + URLEncoder.encode(apiId, java.nio.charset.StandardCharsets.UTF_8);
        Request request = new Request.Builder()
                .url(url)
                .header("Authorization", "Bearer " + bearerToken)
                .header("Accept", "application/json")
                .build();

        try (Response response = client.newCall(request).execute()) {
            if (response.isSuccessful() && response.body() != null) {
                return mapper.readValue(response.body().string(), new TypeReference<Map<String, Object>>() {});
            } else {
                System.err.printf("⚠️ Failed to fetch API metadata %s: %d %s%n", apiId, response.code(), response.message());
            }
        } catch (IOException e) {
            System.err.println("❌ Error fetching API metadata for " + apiId + ": " + e.getMessage());
        }
        return null;
    }

    private static void writeFindingsToJson(List<Map<String, Object>> findings, Path path) {
        try (Writer writer = Files.newBufferedWriter(path)) {
            mapper.writerWithDefaultPrettyPrinter().writeValue(writer, findings);
            System.out.println("✅ JSON written to: " + path.toString());
        } catch (IOException e) {
            System.err.println("⚠️ Failed to write JSON: " + e.getMessage());
        }
    }

    private static void writeFindingsToCsv(List<Map<String, Object>> findings, Path path) {
        if (findings.isEmpty()) {
            System.out.println("⚠️ No findings to write to CSV.");
            return;
        }

        // Flatten each finding into a map of Strings for CSV
        List<Map<String, String>> flatFindings = new ArrayList<>();
        for (Map<String, Object> finding : findings) {
            Map<String, String> flat = new LinkedHashMap<>();
            for (String key : RETURN_FIELDS) {
                Object val = finding.get(key);
                if (val instanceof List) {
                    flat.put(key, String.join(", ", toStringList((List<?>) val)));
                } else if (val instanceof Map) {
                    try {
                        flat.put(key, mapper.writeValueAsString(val));
                    } catch (JsonProcessingException e) {
                        flat.put(key, val.toString());
                    }
                } else {
                    flat.put(key, val == null ? "" : val.toString());
                }
            }
            // Add apiDetails count
            Object apiDetails = finding.get("apiDetails");
            if (apiDetails instanceof List) {
                flat.put("apiDetailsCount", String.valueOf(((List<?>) apiDetails).size()));
            } else {
                flat.put("apiDetailsCount", "0");
            }
            flatFindings.add(flat);
        }

        try (BufferedWriter writer = Files.newBufferedWriter(path);
             CSVPrinter csvPrinter = new CSVPrinter(writer, CSVFormat.DEFAULT
                     .withHeader(concatArrays(RETURN_FIELDS.toArray(new String[0]), new String[]{"apiDetailsCount"})))) {
            for (Map<String, String> record : flatFindings) {
                List<String> row = new ArrayList<>();
                for (String header : csvPrinter.getHeaderNames()) {
                    row.add(record.getOrDefault(header, ""));
                }
                csvPrinter.printRecord(row);
            }
            csvPrinter.flush();
            System.out.println("✅ CSV written to: " + path.toString());
        } catch (IOException e) {
            System.err.println("⚠️ Failed to write CSV: " + e.getMessage());
        }
    }

    private static List<String> toStringList(List<?> list) {
        List<String> out = new ArrayList<>();
        for (Object o : list) {
            out.add(o == null ? "" : o.toString());
        }
        return out;
    }

    @SafeVarargs
    private static <T> T[] concatArrays(T[] first, T[] second) {
        T[] result = Arrays.copyOf(first, first.length + second.length);
        System.arraycopy(second, 0, result, first.length, second.length);
        return result;
    }
}

File: findings.py
#!/usr/bin/env python3
"""
Fetch findings from Noname API, enrich each finding with API metadata (from apiId and relatedApiIds),
then save results to JSON and CSV.

Summary of Your Findings Fetcher Script
Authentication & Token Caching

Authenticates to Noname API using client_id and client_secret.

Caches the Bearer token locally for up to 8 hours to avoid redundant auth calls.

Fetch Findings

Queries the findings API within a specified time window (--start and --end or --hours back).

Requests a comprehensive set of fields (returnFields) for each finding.

Handles pagination with offset and limit.

Retries on network or auth errors with exponential backoff.

API Metadata Enrichment

For each finding, collects its apiId and all relatedApiIds.

Fetches detailed API metadata for each unique API ID.

Adds this API metadata as a list under "apiDetails" inside each finding.

Output Formats

Writes enriched findings with full API metadata to JSON.

Writes a flattened CSV summary containing all requested fields plus a count of API metadata entries per finding.

Properly formats arrays and dicts in CSV for readability.

Logging and Error Handling

Logs token and request activity.

Prints clear error messages for failed requests.

Handles token expiration transparently by refreshing.

Key Benefits
Comprehensive Data: Fetches detailed findings and enriches them with related API details.

Efficiency: Avoids redundant API metadata calls by caching seen API IDs in-memory per run.

Flexible: Accepts time ranges or hours back via CLI arguments.

User-friendly Output: JSON for full detail and CSV for easy spreadsheet analysis.

Robust: Retries on transient errors and handles token expiration automatically.


"""

import argparse
import csv
import json
import logging
import random
import time
from datetime import datetime, timedelta, timezone
from pathlib import Path
from typing import Optional, List, Dict

import requests

# ── Configuration ──────────────────────────────────────────────────────────────
from config import CLIENT_ID, CLIENT_SECRET, HOST, AUTH_URL

TOKEN_CACHE_FILE = Path("/tmp/token_cache.json")
TOKEN_EXPIRY_HOURS = 8

FINDINGS_API = f"{HOST}/api/v4/findings"
CSV_PATH = "/tmp/findings_report.csv"
JSON_PATH = "/tmp/findings_report.json"
LIMIT = 50

# ── Logging ────────────────────────────────────────────────────────────────────
logging.basicConfig(
    filename="/tmp/findings_log.txt",
    level=logging.INFO,
    format="%(asctime)s - %(message)s",
)

RETURN_FIELDS = [
    "id", "title", "url", "typeId", "apiId", "module", "host", "path", "method",
    "resourceGroupName", "status", "severity", "owaspTags", "complianceFrameworkTags",
    "vulnerabilityFrameworkTags", "detectionTime", "lastUpdate", "triggeredOn",
    "description", "impact", "remediation", "investigate", "comments", "tickets",
    "externalTickets", "evidence", "source", "hasRelatedIncidents", "tagsIds",
    "relatedApiIds"
]


def get_cached_token() -> Optional[str]:
    if TOKEN_CACHE_FILE.exists():
        try:
            with TOKEN_CACHE_FILE.open("r") as f:
                data = json.load(f)
            token = data.get("accessToken")
            timestamp = data.get("timestamp")
            if token and timestamp:
                token_time = datetime.fromisoformat(timestamp)
                if datetime.now(timezone.utc) - token_time < timedelta(hours=TOKEN_EXPIRY_HOURS):
                    print("🔐 Using cached token.")
                    return token
        except (json.JSONDecodeError, ValueError):
            print("⚠️ Failed to parse cached token.")
    return None


def cache_token(token: str) -> None:
    with TOKEN_CACHE_FILE.open("w") as f:
        json.dump({
            "accessToken": token,
            "timestamp": datetime.now(timezone.utc).isoformat()
        }, f)


def get_bearer_token(force_refresh: bool = False) -> Optional[str]:
    if not force_refresh:
        cached = get_cached_token()
        if cached:
            return cached

    payload = {
        "client_id": CLIENT_ID,
        "client_secret": CLIENT_SECRET
    }

    try:
        response = requests.post(AUTH_URL, json=payload)
        if response.status_code in [200, 201]:
            token = response.json().get("accessToken")
            if token:
                cache_token(token)
                return token
        print(f"❌ Token fetch failed: {response.status_code} - {response.text}")
    except requests.exceptions.RequestException as e:
        print(f"❌ Auth error: {e}")
    return None


def build_headers(token: str) -> Dict[str, str]:
    return {
        "Authorization": f"Bearer {token}",
        "Accept": "application/json",
        "Content-Type": "application/json"
    }


def retry_request(method: str, url: str, headers: dict, params: Optional[dict] = None, retries: int = 5) -> Optional[requests.Response]:
    global BEARER_TOKEN
    for attempt in range(retries):
        try:
            response = getattr(requests, method)(url, headers=headers, params=params)
            if response.status_code == 200:
                return response
            elif response.status_code == 401:
                print("🔁 Token expired. Refreshing.")
                BEARER_TOKEN = get_bearer_token(force_refresh=True)
                headers["Authorization"] = f"Bearer {BEARER_TOKEN}"
                time.sleep(1.5)
                continue
            else:
                print(f"❗ Request failed ({response.status_code}): {response.text}")
        except requests.exceptions.RequestException as e:
            print(f"❌ Network error: {e}")
        time.sleep((2 ** attempt) + random.uniform(0, 1))
    return None


def get_time_window(start_arg: Optional[str], end_arg: Optional[str], hours_back: int = 24):
    now = datetime.now(timezone.utc)
    if start_arg and end_arg:
        return start_arg, end_arg
    end_time = now
    start_time = now - timedelta(hours=hours_back)
    return start_time.isoformat(), end_time.isoformat()


def fetch_findings(start_time: str, end_time: str) -> List[dict]:
    findings = []
    offset = 0

    while True:
        params = {
            "sortDesc": "true",
            "limit": LIMIT,
            "offset": offset,
            "detectionStartDate": start_time,
            "detectionEndDate": end_time,
            "lastUpdateStartDate": start_time,
            "lastUpdateEndDate": end_time,
            "returnFields": RETURN_FIELDS
        }

        response = retry_request("get", FINDINGS_API, build_headers(BEARER_TOKEN), params=params)
        if not response:
            break

        data = response.json()
        entities = data.get("entities", [])
        findings.extend(entities)
        print(f"📥 Retrieved {len(entities)} findings (offset {offset})")
        if not data.get("moreEntities"):
            break
        offset += LIMIT
    return findings


def fetch_api_metadata(api_id: str, token: str) -> Optional[dict]:
    url = f"{HOST}/api/v3/apis/{api_id}"
    headers = build_headers(token)
    try:
        response = requests.get(url, headers=headers)
        if response.status_code == 200:
            return response.json()
        else:
            print(f"⚠️ Failed to fetch API {api_id}: {response.status_code}")
    except requests.exceptions.RequestException as e:
        print(f"❌ API fetch error for {api_id}: {e}")
    return None


def stringify(value):
    if isinstance(value, list):
        return ", ".join(str(v) for v in value)
    elif isinstance(value, dict):
        return json.dumps(value)
    return value


def flatten_finding(finding: dict) -> dict:
    # Flatten all RETURN_FIELDS + flatten apiDetails summary (just API ids count)
    flat = {field: stringify(finding.get(field)) for field in RETURN_FIELDS}
    # Add a summary field for how many APIs detailed
    if "apiDetails" in finding:
        flat["apiDetailsCount"] = len(finding["apiDetails"])
    else:
        flat["apiDetailsCount"] = 0
    return flat


def write_findings_to_csv(findings: List[dict], output_path: str = CSV_PATH) -> None:
    if not findings:
        print("⚠️ No findings to write to CSV.")
        return

    flattened = [flatten_finding(f) for f in findings]
    keys = RETURN_FIELDS + ["apiDetailsCount"]

    with open(output_path, "w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=keys, extrasaction="ignore")
        writer.writeheader()
        writer.writerows(flattened)

    print(f"✅ CSV written to: {output_path}")


def write_findings_to_json(findings: List[dict], output_path: str = JSON_PATH) -> None:
    if not findings:
        print("⚠️ No findings to write to JSON.")
        return

    with open(output_path, "w", encoding="utf-8") as f:
        json.dump(findings, f, indent=2)

    print(f"✅ JSON written to: {output_path}")


# --- MAIN ---


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Fetch findings from Noname API")
    parser.add_argument("--start", help="Start ISO timestamp (UTC)", default=None)
    parser.add_argument("--end", help="End ISO timestamp (UTC)", default=None)
    parser.add_argument("--hours", help="Hours back if no start/end", type=int, default=24)
    args = parser.parse_args()

    BEARER_TOKEN = get_bearer_token()
    if not BEARER_TOKEN:
        raise SystemExit("❌ Failed to authenticate.")

    start, end = get_time_window(args.start, args.end, args.hours)
    print(f"📅 Fetching findings from: {start} to {end}")

    raw_findings = fetch_findings(start, end)

    # Enrich findings with API metadata
    enriched_findings = []
    seen_api_ids = set()
    for finding in raw_findings:
        api_ids = set()
        if finding.get("apiId"):
            api_ids.add(finding["apiId"])
        for rid in finding.get("relatedApiIds", []):
            api_ids.add(rid)

        api_details = []
        for api_id in api_ids:
            if api_id not in seen_api_ids:
                metadata = fetch_api_metadata(api_id, BEARER_TOKEN)
                if metadata:
                    api_details.append(metadata)
                seen_api_ids.add(api_id)
        finding["apiDetails"] = api_details
        enriched_findings.append(finding)

    write_findings_to_csv(enriched_findings)
    write_findings_to_json(enriched_findings)
    print(f"📊 Total findings: {len(enriched_findings)}")
